File: benchmark-cpp/src/main/cpp/HeapSortTest.cpp
#include <time.h>
#include <iostream>
#include <vector>
#include <math.h>
#include <algorithm>

using namespace std;



class HeapSortTest
{
	public:

	/**
     * NB_TESTS
     */
	static const int NB_TESTS                = 10;

    /**
     * NB_OF_EXCLUSION_MIN_MAX
     */
    static const int NB_OF_EXCLUSION_MIN_MAX = 0;

    /**
     * SIZE
     */
    static const int SIZE                    = 5000000;
    
    /**
     * NB_HEAPSORT_TESTS
     */
    static const int NB_HEAPSORT_TESTS       = 1;


    /**
     * averageTimeWithoutMinMax.
     * 
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    static long averageTimeWithoutMinMax(vector<long> executionTimes, int numberOfMinMaxToRemove)
    {
        long average = 0;
	int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0;
	if (minMax != 0)	
		sort(executionTimes.begin(), executionTimes.end());
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
			average += executionTimes[i];
        }
        average = average / (executionTimes.size() - minMax * 2);
        return average;
    }
	
    
	/**
     * relativeDeviationTime.
     *
     * @param executionTimes
     * @return
     */
    static double relativeDeviationTimeWithoutMinMax(vector<long> executionTimes, int numberOfMinMaxToRemove)
    {
	long averageTime = HeapSortTest::averageTimeWithoutMinMax(executionTimes, numberOfMinMaxToRemove);
        double deviation = 0;
	int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0;
	if (minMax != 0)	
		sort(executionTimes.begin(), executionTimes.end());
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            deviation += pow(executionTimes[i] - averageTime, 2.0);
        }
        return (deviation == 0) ? 0.0 : ((int) ((sqrt((double)(deviation / executionTimes.size())) / averageTime * 100) * 100)) / 100.0;
    }	
	#define IM 139968
	#define IA   3877
	#define IC  29573


	static double gen_random(double max) {
		static long last = 42;
		return( max * (last = (last * IA + IC) % IM) / IM );
	}

	
	/**
     * heapsort.
     *
     */
	static void heapsort(int n, double *ra) 
	{
		int i, j;
		int ir = n;
		int l = (n >> 1) + 1;
		double rra;

		for (;;) 
		{
			if (l > 1) 
			{
				rra = ra[--l];
			} 
			else 
			{
				rra = ra[ir];
				ra[ir] = ra[1];
				if (--ir == 1) 
				{
				ra[1] = rra;
				return;
				}
			}
			i = l;
			j = l << 1;
			while (j <= ir) 
			{
				if (j < ir && ra[j] < ra[j+1]) 
				{ 
					++j; 
				}
				if (rra < ra[j]) 
				{
					ra[i] = ra[j];
					j += (i = j);
				} 
				else 
				{
					j = ir + 1;
				}
			}
			ra[i] = rra;
		}
	}
	
	/**
     * testCreationOfClassWith1String
     */
    static long testHeapSort()
    {
		
		double * ary;
		ary = new double[SIZE+1];
		for (int i = SIZE; i != 0; i--) 
		{
			ary[i] = gen_random(1.0);
		}
		clock_t start = clock();
		for (int i = NB_HEAPSORT_TESTS; i != 0; i--)
        {
			HeapSortTest::heapsort(SIZE, ary);
		}
        clock_t end = clock();
		
        long executionTime = (long) ((end - start) / ((double) CLOCKS_PER_SEC) * 1000.0);
		cout << "[HeapSortTest], Sort an array of double " << SIZE  << ",, snapshot time," << executionTime << "\n";
        return executionTime;
    }

	static int main (int size, char** args)
	{
		int nbTests = HeapSortTest::NB_TESTS;
		int nbOfExclusionMinMax = HeapSortTest::NB_OF_EXCLUSION_MIN_MAX;
	    
		// TODO does not work
		//if (args && size > 0)
		//{
		//    nbTests = atoi(args[0].c_str());
		//}
		//if (args && size > 1)
		//{
		//    nbOfExclusionMinMax = atoi(args[1].c_str());
		//}
		
		vector<long> executionTimes;
		
		executionTimes.clear();
		for (int i = nbTests; i != 0; i--)	
			executionTimes.push_back
				(HeapSortTest::testHeapSort());
		std::sort(executionTimes.begin(), executionTimes.end());
		cout << "[HeapSortTest], Sort an array of double " << HeapSortTest::SIZE << ",, average time,"<< HeapSortTest::averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) << ", min time," << executionTimes[0] << ", max time," << executionTimes[executionTimes.size() - 1] << ", relative deviation time," << HeapSortTest::relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) << "\n";

		return 0;
	}
};





/*
int main(string* args, int size)
{
	HeapSortTest::main(args, size);
}
*/

File: benchmark-java/src/main/java/HeapSortTest.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * HeapSortTest.
 * 
 * @author Alexandre
 * 
 */
public final class HeapSortTest
{
    /**
     * NB_TESTS
     */
    private static final int NB_TESTS                = 5;
    
    /**
     * NB_OF_EXCLUSION_MIN_MAX
     */
    private static final int NB_OF_EXCLUSION_MIN_MAX = 2;
    
    /**
     * SIZE
     */
    private static final int SIZE                    = 5 * 1000 * 1000;
    
    /**
     * NB_HEAPSORT_TESTS
     */
    private static final int NB_HEAPSORT_TESTS       = 1;

    
    /**
     * averageTimeWithoutMinMax.
     * 
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    @SuppressWarnings("boxing")
    public static long averageTimeWithoutMinMax(final List<Long> executionTimes, final int numberOfMinMaxToRemove)
    {
        long average = 0;
        final int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0; 
        if (minMax != 0)
        {
            Collections.sort(executionTimes);
        }
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            average += executionTimes.get(i);
        }
        average = average / (executionTimes.size() - minMax * 2);
        return average;
    }
    
    
    /**
     * relativeDeviationTimeWithoutMinMax.
     *
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    @SuppressWarnings("boxing")
    public static double relativeDeviationTimeWithoutMinMax(final List<Long> executionTimes, final int numberOfMinMaxToRemove)
    {
        final long averageTimeWithoutMinMax = averageTimeWithoutMinMax(executionTimes, numberOfMinMaxToRemove);
        long deviation = 0;
        final int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0; 
        if (minMax != 0)
        {
            Collections.sort(executionTimes);
        }
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            deviation += Math.pow(executionTimes.get(i) - averageTimeWithoutMinMax, 2);
        }
        return ((int) ((Math.sqrt(deviation / (executionTimes.size() - 2 * minMax)) / averageTimeWithoutMinMax * 100) * 100)) / 100d;
    }
    

    
    /**
     * testHeapSort.
     * 
     * @return
     */
    public static long testHeapSort()
    {
        double[] array = new double[SIZE];
        for (int i = SIZE; i-- != 0;)
            array[i] = RandomizerHeapSort.getRandomValue(1);
        long start = System.currentTimeMillis();
        start = System.currentTimeMillis();
        for (int i = NB_HEAPSORT_TESTS; i != 0; i--)
        {
            heapSort(array);
            
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[HeapSortTest], Sort an array of double " + SIZE + ",, snapshot time," + executionTime);
        return executionTime;
    }
    
    /**
     * heapsort.
     * 
     * @param n
     * @param ra
     */
    public static void heapSort(double[] ra)
    {
        int l, ir;
        double rra;
        
        l = ((ra.length - 1) >> 1) + 1;
        ir = ra.length - 1;
        for (;;)
        {
            if (l > 1)
                rra = ra[--l];
            else
            {
                rra = ra[ir];
                ra[ir] = ra[1];
                if (--ir == 1)
                {
                    ra[1] = rra;
                    return;
                }
            }
            
            int i = l;
            int j = l << 1;
            
            while (j <= ir)
            {
                
                if (j < ir && ra[j] < ra[j + 1])
                    ++j;
                
                if (rra < ra[j])
                {
                    ra[i] = ra[j];
                    j += (i = j);
                }
                else
                    j = ir + 1;
            }
            
            ra[i] = rra;
        }
    }
    
    /**
     * main.
     * 
     * @param args
     */
    @SuppressWarnings("boxing")
    public static void main(final String[] args)
    {
        final int nbTests = (args != null && args.length >= 1 && args[0] != null) ?  Integer.valueOf(args[0]): NB_TESTS;
        final int nbOfExclusionMinMax = (args != null && args.length >= 2 && args[1] != null) ? Integer.valueOf(args[1]) : NB_OF_EXCLUSION_MIN_MAX;
        
        final List<Long> executionTimes = new ArrayList<Long>(nbTests);
        
        for (int i = nbTests; i != 0; i--)
            executionTimes.add(HeapSortTest.testHeapSort());
        System.out.println("[HeapSortTest], Sort an array of double " + SIZE + ",, average time," + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time,"
                + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes) + ", relative deviation time,"
                + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();
        
    }
    
}

/**
 * Randomizer.
 * 
 * @author Alexandre
 * 
 */
final class RandomizerHeapSort
{
    private static long lastRandom = 42;
    
    /**
     * Constructor.
     * 
     */
    private RandomizerHeapSort()
    {
        //
    }
    
    /**
     * resetRandomizer.
     * 
     */
    public static void resetRandomizer()
    {
        lastRandom = 42;
    }
    
    /**
     * getRandomValue.
     * 
     * @param max
     * @return
     */
    public static double getRandomValue(final double max)
    {
        return (max * (lastRandom = (lastRandom * 3877 + 29573) % 139968) / 139968);
    }
}
